feat: Geothermal Gradient for Single Phase flow

npillardou opened the pull request:
The purpose of this PR is to add the geothermal gradient capacities for single-phase flow.

At the moment the temperature VS Elevation is correlated with the composition VS Elevation, maybe this point could be extended to something that is independent of the composition: 

    <HydrostaticEquilibrium
      name="Initialisation"
      datumElevation="-3790"
      datumPressure="4.3e7"
      objectPath="ElementRegions"
      initialPhaseName="water"
      componentNames="{ co2, water }"
      componentFractionVsElevationTableNames="{ CO2VD, H2OVD }"
      temperatureVsElevationTableName="initTempTable" />  


Further work is needed to account for composition in the hydrostatic pressure kernel, as it is done in CompositionalMultiphaseFlow.

npillardou committed:
Integrate SinglePhase Geothermal Gradient capacities

paveltomin committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -505,6 +505,21 @@ void SinglePhaseBase::computeHydrostaticEquilibrium( DomainPartition & domain )
+      tempTableGradient = true;
+      string const tempTableName = fs.getTemperatureVsElevationTableName();
+      TableFunction const & tempTable = functionManager.getGroup< TableFunction >( tempTableName );
+      tempTableWrapper = tempTable.createKernelWrapper();
```
this should be used in `HydrostaticPressureKernel`, no? i mean let's say density formally depends on temperature, then temperature should be used there

npillardou replied:
Yes I agree with you, we should modify the kernel into something like: 

HydrostaticPressureKernel::launch( numPointsInTable,
                                           maxNumEquilIterations,
                                           equilTolerance,
                                           gravVector,
                                           minElevation,
                                           elevationIncrement,
                                           datumElevation,
                                           datumPressure,
                                           fluidWrapper,
                                           tempTableWrapper,
                                           elevationValues.toNestedView(),
                                           pressureValues.toView() ); 
                                           
But behind that, it requires further developments in the constitutive also

paveltomin replied:
> But behind that, it requires further developments in the constitutive also

can you elaborate?

jhuang2601 replied:
> this should be used in `HydrostaticPressureKernel`, no? i mean let's say density formally depends on temperature, then temperature should be used there

Yes, add temperature field in `HydrostaticPressureKernel`, considering temperature dependent fluid density

npillardou committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

tjb-ltk commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -562,6 +577,7 @@ void SinglePhaseBase::computeHydrostaticEquilibrium( DomainPartition & domain )
+    arrayView1d< real64 > const temp = subRegion.getField< fields::flow::temperature >();
 
     RAJA::ReduceMin< parallelDeviceReduce, real64 > minPressure( LvArray::NumericLimits< real64 >::max );
 
```
You could add.   kernelLaunchSelectorThermalSwitch and then use constexpr to have compile time inclusion of thermal code instead of if tempTableGradient

jafranc replied:
I think the dispatch will only be valid if we are actually dealing with thermal dependent solver (as the `m_isThermal` in the class below) and not if the intent is to impose a passive temperature field

jhuang2601 replied:
Just switch to `m_isThermal`

paveltomin commented:
@npillardou are you planning to finish this PR or it can be closed?

herve-gross commented:
@dkachuma and @jhuang2601 please take a look

npillardou commented:
> @npillardou are you planning to finish this PR or it can be closed?

Sorry @paveltomin I totally forgot to answer you. I should take some time to investigate a good implementation, maybe with the help of Jian and Dick.

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
use full exp for densRelation in single phase flow

jhuang2601 committed:
use temperature in HydrostaticPressureKernel

jafranc reviewed (commented):
Would it be possible to make it consistent with this:
https://github.com/GEOS-DEV/GEOS/blob/45f272c3d8313d6443df260436bc1a2cf5244b6b/src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/HydrostaticPressureKernel.hpp#L49-L56

from #3795

jafranc reviewed (commented):
@npillardou can you add an inputFile using it ?

jhuang2601 committed:
update SinglePhaseBase

jhuang2601 committed:
update SingleFluidBase

jhuang2601 committed:
fix DirichletFluxComputeKernel

jafranc commented:
Ok looking at it I am a bit puzzled, most likely because I am a little rusted.

IIUC we are trying to create a shorthand to have a single phase non thermally dependent case to initialize a _passive_ temperature field base on a table under the `<HydrostaticEquilibrium>` tag, am I correct @npillardou  ?

jafranc commented:
If *true*, I would group everything temperature introduced in `SinglePhaseBase::computeHydrostaticEquilibrium()` at the end of the function, to the expense of repeating the loops (IIRC the cost won't be much).

If *false* and there is plan and intention to have the single phase initialization to take into account both pressure and temperature, then as @tjb-ltk we need some `constexpr` dispatch that could elude code generation in the part it is not needed and error thrown if there is thermal simulation launched without thermal init (isn't it ?)

jhuang2601 commented:
> If _true_, I would group everything temperature introduced in `SinglePhaseBase::computeHydrostaticEquilibrium()` at the end of the function, to the expense of repeating the loops (IIRC the cost won't be much).
> 
> If _false_ and there is plan and intention to have the single phase initialization to take into account both pressure and temperature, then as @tjb-ltk we need some `constexpr` dispatch that could elude code generation in the part it is not needed and error thrown if there is thermal simulation launched without thermal init (isn't it ?)

@jafranc It should be the latter one. As of now, I just address Pavel's comment to use temperature field in `HydrostaticPressureKernel`. This is still a WIP process to add `constexpr` dispatch.

jhuang2601 committed:
seperate HydrostaticPressureKernel for thermal and isothermal cases

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
include ThermalHydrostaticPressureKernel

jhuang2601 committed:
update ThermalHydrostaticPressureKernel

jhuang2601 committed:
code style

jhuang2601 committed:
code style

jhuang2601 committed:
missing semicolon

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/kernels/singlePhase/ThermalHydrostaticPressureKernel.hpp:
```diff
@@ -0,0 +1,223 @@
+                              real64 const & equilTolerance,
+                              real64 const (&gravVector)[ 3 ],
+                              FLUID_WRAPPER fluidWrapper,
+                              TableFunction::KernelWrapper tempTableWrapper,
```
Shouldn't temperature value be passed as a `real64 const`/`real64` to be consistent with unitarian nature of the intended interface ?
```suggestion
                              real64 const & temp,
```

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/kernels/singlePhase/ThermalHydrostaticPressureKernel.hpp:
```diff
@@ -0,0 +1,223 @@
+    // Step 1: guess the pressure with the refDensity
+
+    real64 const gravCoef = gravVector[2] * ( refElevation - newElevation );
+    real64 const temp = tempTableWrapper.compute( &newElevation );
```
```suggestion
```

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/kernels/singlePhase/ThermalHydrostaticPressureKernel.hpp:
```diff
@@ -0,0 +1,223 @@
+                                  equilTolerance,
+                                  gravVector,
+                                  fluidWrapper,
+                                  tempTableWrapper,
```
```suggestion
                                 temp[iRef], 
```

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/kernels/singlePhase/ThermalHydrostaticPressureKernel.hpp:
```diff
@@ -0,0 +1,223 @@
+                                    equilTolerance,
+                                    gravVector,
+                                    fluidWrapper,
+                                    tempTableWrapper,
```
```suggestion
                                    temp[iRef+i],
```

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/kernels/singlePhase/ThermalHydrostaticPressureKernel.hpp:
```diff
@@ -0,0 +1,223 @@
+                                    equilTolerance,
+                                    gravVector,
+                                    fluidWrapper,
+                                    tempTableWrapper,
```
```suggestion
                                    temp[iRef-i],
```

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -518,28 +536,54 @@ void SinglePhaseBase::computeHydrostaticEquilibrium( DomainPartition & domain )
+                                               datumElevation,
+                                               datumPressure,
+                                               fluidWrapper,
+                                               tempTableWrapper,
```
```suggestion
                                               tempValues.toView(),
```

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -492,6 +500,16 @@ void SinglePhaseBase::computeHydrostaticEquilibrium( DomainPartition & domain )
     // Step 3.2: retrieve the fluid model to compute densities
     // we end up with the same issue as in applyDirichletBC: there is not a clean way to retrieve the fluid info
 
```
Why not group that closer to the call-site ?

jafranc reviewed (commented):
Except from a small comment on the interface, to pass simple `const/non-const  real64` instead of full table wrapper and group all new addition at the bottom instead of across the file, the implementation looks good to me. Thanks @jhuang2601 !

Maybe we should also touch-base the problem of having to provid obvious compositionVSDepth in single phase case instead of automatically fill them with 1-or-0 ?

jafranc committed:
Getting rid of component in HSE init for singlephase flow

npillardou committed:
Add Geothermal gradient test + add in automatic integrated tests

npillardou commented:
To keep track : 

Based on the previous development, I have just added a 3D test with a geothermal gradient option. 
I included this case in the integrated tests. 

However, I noticed that the 'same' case in isothermal conditions, which was already present in the folder, was no longer converging. Therefore, I modified the lineSearch option to 'None' to resolve this issue. I also added this case to the integrated tests — it might be interesting to have it here too.

jhuang2601 committed:
update ats for steps

paveltomin committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
add temperatureVsElevationTableName for a thermal simulation

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
switch to direct solver for smoke tests

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 commented:
@MelReyCG @arng40 For the PR, this CI check kept failing on `testFlowStatistics` :https://github.com/GEOS-DEV/GEOS/actions/runs/18139901718/job/51627947541?pr=3193
Can you take a look and advise how to fix it?

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

MelReyCG commented:
This does not seem really like a issue really related with the test but a memory corruption.
More precisely, it seems like a corrupted memory detection while freeing the memory of a CellElementSubregion .

First, I would suggest testing that with the problematic system with bounds-checking.
Then testing to remove parts of the PR, to try to see the problem source.

jafranc committed:
Update ci_tests.yml - Bound check cuda+clang

paveltomin committed:
Disable GEOS bounds checking in CI tests

paveltomin committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
update .integrated_tests.yaml

jhuang2601 committed:
update BASELINE_NOTES.md

jhuang2601 commented:
Due to change in `CompressibleSinglePhaseUpdate< ExponentApproximationType::Full`, all smoke tests with `CompressibleSinglePhaseFluid` failed as expected and hence rebaseline.

paveltomin committed:
Disable sccache usage in CI workflow

paveltomin committed:
Remove USE_SCCACHE variable from CI workflow

paveltomin committed:
Remove USE_SCCACHE from CI tests workflow

Removed USE_SCCACHE variable from CI configuration.

paveltomin committed:
Change KernelWrapper to use Linear approximation type

rrsettgast commented on src/coreComponents/constitutive/fluid/singlefluid/CompressibleSinglePhaseFluid.hpp:
```diff
@@ -155,6 +155,7 @@ class CompressibleSinglePhaseFluid : public SingleFluidBase
 
   /// Type of kernel wrapper for in-kernel update (TODO: support multiple EAT, not just linear)
   using KernelWrapper = CompressibleSinglePhaseUpdate< ExponentApproximationType::Linear, ExponentApproximationType::Linear >;
+//  using KernelWrapper = CompressibleSinglePhaseUpdate< ExponentApproximationType::Full, ExponentApproximationType::Linear >;
```
```suggestion
```

paveltomin committed:
Change KernelWrapper type to Full approximation

paveltomin committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

Commit:
try this

Commit:
Merge branch 'feature/npillardou/singlePhaseGeothermalGradient' of https://github.com/GEOS-DEV/GEOS into feature/npillardou/singlePhaseGeothermalGradient

Commit:
try that

paveltomin commented on src/coreComponents/constitutive/fluid/singlefluid/CompressibleSinglePhaseFluid.cpp:
```diff
@@ -110,9 +110,9 @@ void CompressibleSinglePhaseFluid::postInputInitialization()
-    GEOS_THROW_IF_NE_MSG( value, ExponentApproximationType::Linear,
-                          GEOS_FMT( "{}: invalid model type in attribute '{}' (only linear currently supported)", getFullName(), attribute ),
-                          InputError );
+    GEOS_THROW_IF( value != ExponentApproximationType::Linear && value != ExponentApproximationType::Full,
```
linear is not supported anymore?

jhuang2601 replied:
Both linear and full are valid. If the value is not one of these, it throws an error.

paveltomin committed:
Remove default move assignment operator from KernelWrapper

Removed default move assignment operator and its implementation from KernelWrapper.

paveltomin committed:
Remove blank line in KernelWrapper definition

Removed unnecessary blank line in KernelWrapper class.

paveltomin committed:
Remove blank line in KernelWrapper definition

Removed unnecessary blank line in KernelWrapper class.

Commit:
fix?

Commit:
correct check

Commit:
correct check

Commit:
try this

Commit:
revert

Commit:
debug output

Commit:
Merge remote-tracking branch 'origin/develop' into feature/npillardou/singlePhaseGeothermalGradient

Commit:
make host

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -552,17 +599,25 @@ void SinglePhaseBase::computeHydrostaticEquilibrium( DomainPartition & domain )
-    forAll< parallelDevicePolicy< > >( targetSet.size(), [=] GEOS_HOST_DEVICE ( localIndex const i )
+    RAJA::ReduceMin< parallelHostReduce, real64 > minPressure( LvArray::NumericLimits< real64 >::max );
+
+    forAll< parallelHostPolicy >( targetSet.size(), [=] GEOS_HOST_DEVICE ( localIndex const i )
```
@rrsettgast @klevzoff 
changing this from `parallelDevicePolicy` to `parallelHostPolicy` seems to resolve that unit test failure

jhuang2601 committed:
remove debug output

jhuang2601 committed:
update .integrated_tests.yaml

jhuang2601 committed:
update BASELINE_NOTES.md

jhuang2601 committed:
update .integrated_tests.yaml

jhuang2601 committed:
Merge branch 'develop' into feature/npillardou/singlePhaseGeothermalGradient

jhuang2601 committed:
fix the timeout issue of PoroElastic_hybridHexPrism_co2_fim_3d

jhuang2601 committed:
revert file path

jhuang2601 committed:
update .integrated_tests.yaml

jhuang2601 committed:
update BASELINE_NOTES.md
